internal/operations/operation_updater.go

```diff
@@ -0,0 +1,342 @@
+			if batch == nil {
+				batch = &operationUpdaterBatch{}
+				batch.timeoutContext, batch.timeoutCancel = context.WithTimeout(ctx, ou.conf.batchTimeout)
+			}
+			batch.updates = append(batch.updates, work)
+		case <-timeoutContext.Done():
+			timedOut = true
+		}
+
+		if batch != nil && (timedOut || len(batch.updates) >= ou.conf.maxInserts) {
+			batch.timeoutCancel()
+			err := ou.retry.Do(ctx, "operation batch update", func(attempt int) (retry bool, err error) {
+				return true, ou.database.RunAsGroup(ctx, func(ctx context.Context) error {
+					return ou.doBatchUpdate(ctx, batch.updates)
+				})
+			})
+			if err != nil {
+				log.L(ctx).Debugf("Operation update worker exiting: %s", err)
+				return
+			}
+			batch = nil
+		}
+	}
+}
+
+func (ou *operationUpdater) doBatchUpdate(ctx context.Context, updates []*OperationUpdate) error {
+
+	// Get all the operations that match
+	opIDs := make([]driver.Value, len(updates))
+	for idx, update := range updates {
+		opIDs[idx] = update.ID
+	}
+	opFilter := database.OperationQueryFactory.NewFilter(ctx).In("id", opIDs)
+	ops, _, err := ou.database.GetOperations(ctx, opFilter)
+	if err != nil {
+		return err
+	}
+
+	// Get all the transactions for these operations
+	txIDs := make([]driver.Value, 0, len(ops))
+	for _, op := range ops {
+		if op.Transaction != nil {
+			txIDs = append(txIDs, op.Transaction)
+		}
+	}
+	var transactions []*fftypes.Transaction
+	if len(txIDs) > 0 {
+		txFilter := database.TransactionQueryFactory.NewFilter(ctx).In("id", txIDs)
+		transactions, _, err = ou.database.GetTransactions(ctx, txFilter)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Spin through each update seeing what DB updates we need to do
+	for _, update := range updates {
+		if err := ou.doUpdate(ctx, update, ops, transactions); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (ou *operationUpdater) doUpdate(ctx context.Context, update *OperationUpdate, ops []*fftypes.Operation, transactions []*fftypes.Transaction) error {
+
+	// Find the operation we already retrieved, and do the update
+	var op *fftypes.Operation
+	for _, candidate := range ops {
+		if update.ID.Equals(candidate.ID) {
+			op = candidate
+			break
+		}
+	}
+	if op == nil {
+		log.L(ctx).Warnf("Operation update '%s' ignored, as it was not submitted by this node", update.ID)
+		return nil
+	}
+
+	// Match a TX we already retireved, if found add a specified Blockchain Transaction ID to it
+	var tx *fftypes.Transaction
+	if op.Transaction != nil && update.BlockchainTXID != "" {
+		for _, candidate := range transactions {
+			if op.Transaction.Equals(candidate.ID) {
+				tx = candidate
+				break
+			}
+		}
+	}
+	if tx != nil {
+		if err := ou.txHelper.AddBlockchainTX(ctx, tx, update.BlockchainTXID); err != nil {
+			return err
+		}
+	}
+
+	// Special handling for OpTypeTokenTransfer, which writes an event when it fails
+	if op.Type == fftypes.OpTypeTokenTransfer && update.State == fftypes.OpStatusFailed {
```

Comment:
Quite possibly these "special" cases should be farmed out to registered handlers as a new type of callback. Doesn't need to be done now, but I think we're a step closer to making it possible since the Operations Manager has a map of op types to handlers.

Reply:
Agreed 👍 ... and deferring for now